src/ray/gcs/gcs_server/gcs_placement_group_scheduler.cc

```diff
@@ -33,17 +33,45 @@ GcsPlacementGroupScheduler::GcsPlacementGroupScheduler(
   scheduler_strategies_.push_back(std::make_shared<GcsSpreadStrategy>());
 }
 
-/// This is an initial algorithm to respect pack algorithm.
-/// In this algorithm, we try to pack all the bundle in the first node
-/// and don't care the real resource.
+/// In this algorithm, we try to pack all the bundles in the node which satisfies the
+/// resource requirements and has the least number of bundles.
+/// TODO(ffbin): At present, only one node will be scheduled. If one node does not have
+/// enough resources, we need to divide bundles to multiple nodes. We will implement
+/// it in the next pr.
 ScheduleMap GcsPackStrategy::Schedule(
     std::vector<std::shared_ptr<ray::BundleSpecification>> &bundles,
-    const GcsNodeManager &node_manager) {
+    const std::unique_ptr<ScheduleContext> &context) {
+  // Aggregate required resources.
+  std::unordered_map<std::string, double> required_resources;
+  for (const auto &bundle : bundles) {
+    const auto &resources = bundle->GetRequiredResources().GetResourceMap();
+    for (const auto &iter : resources) {
+      required_resources[iter.first] += iter.second;
+    }
+  }
+
+  // Filter candidate nodes.
+  const auto &alive_nodes = context->node_manager_.GetClusterRealtimeResources();
+  std::vector<std::pair<int64_t, ClientID>> candidate_nodes;
+  for (auto &node : alive_nodes) {
+    if (node.second->IsSubset(required_resources)) {
```

Comment:
Yeah, this `required_resources `contain the aggregation of all the bundle spec. This algorithm is complex(Pack should  avoid excessive resource fragmentation), so i want to write a design document and discuss it with you later. In this PR, I only made a simple implementation.

Reply:
Why don't we iterate through all alive nodes, check if each node satisfies the bundle placement requirement, and put them in there if it does? This will have O(BN) where N: number of nodes & B: number of bundles. -----sorry, I am not sure about this. Can you explain it again, thanks.

Reply:
Can you elaborate a little bit about what it means by strict and soft in this context? Does strict mean, we will schedule placement group only when we can put every bundle into a single node?

Reply:
>Why don't we iterate through all alive nodes, check if each node satisfies the bundle placement requirement, and put them in there if it does? This will have O(BN) where N: number of nodes & B: number of bundles.

I guess this is naive algorithm for soft pack strategy if I understand your strict & soft definition correctly. I just meant this 
```python3
for node in alive_nodes:
    for bundle in bundles:
        if bundle.resource.subset_of(node.resource):
            node.reserve(bundle)
            node.resource -= bundle.resource
```

Reply:
> Can you elaborate a little bit about what it means by strict and soft in this context? Does strict mean, we will schedule placement group only when we can put every bundle into a single node?

The strict pack policy is that all bundles must be scheduled to one node. The soft pack strategy is to schedule all bundles to one node as much as possible. If one node cannot be scheduled, it can also be scheduled to multiple nodes.

Reply:
> > Why don't we iterate through all alive nodes, check if each node satisfies the bundle placement requirement, and put them in there if it does? This will have O(BN) where N: number of nodes & B: number of bundles.
> 
> I guess this is naive algorithm for soft pack strategy if I understand your strict & soft definition correctly. I just meant this
> 
> ```python
> for node in alive_nodes:
>     for bundle in bundles:
>         if bundle.resource.subset_of(node.resource):
>             node.reserve(bundle)
>             node.resource -= bundle.resource
> ```

Thanks